WordPress/gutenberg
# fix: Avoid iOS block appender focus loop

## Related PRs

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5215

## What?

During programmatic Aztec focus events, only update the internal focused
element ref found within `TextInputState` and do not call the native focus
methods. 

## Why?

Fixes wordpress-mobile/WordPress-iOS#18783. Quickly adding multiple rich text
blocks on iOS resulted in a focus loop occurring between the rich text blocks.
The app would become unresponsive and eventually crash.

## How?

The focus callback triggered by Aztec-based programmatic focus events
can result in focus loops between rich text elements.

Android: This intentional no-op function prevents focus loops
originating when the native Aztec module programmatically focuses the
instance. The no-op is explicitly passed as an `onFocus` prop to avoid
future prop spreading from inadvertently introducing focus loops. The
user-facing focus of the element is handled by `onPress` instead.

See: https://github.com/wordpress-mobile/gutenberg-mobile/issues/302

iOS: Programmatic focus from the native Aztec module is required to
ensure the React-based `TextStateInput` ref is properly set when focus
is *returned* to an instance, e.g. dismissing a bottom sheet. If the ref
is not updated, attempts to dismiss the keyboard via the `ToolbarButton`
will fail.

See: https://github.com/wordpress-mobile/gutenberg-mobile/issues/702

The Android keyboard is, likely erroneously, already dismissed in the
contexts where programmatic focus may be required on iOS.

- https://github.com/WordPress/gutenberg/issues/28748
- https://github.com/WordPress/gutenberg/issues/29048
- https://github.com/wordpress-mobile/WordPress-Android/issues/16167

Programmatically swapping input focus creates an infinite loop if the
user taps a different input in between the programmatic focus and
the resulting update to the React Native TextInputState focused element
ref. To mitigate this, the below updates the focused element ref, but
does not call the native focus methods.

See: https://github.com/wordpress-mobile/WordPress-iOS/issues/18783

## Testing Instructions

<details><summary>Verify block appender loop fix</summary>

See wordpress-mobile/WordPress-iOS#18783.

1. Open a post in the editor.
1. Scroll to the bottom of the content.
1. Rapidly tap beneath the last block, which should add an empty Paragraph block.

</details>

<details><summary>Verify link picker keyboard dismissal</summary>

See https://github.com/wordpress-mobile/WordPress-iOS/issues/18783.

### iOS

1. Select a Paragraph block.
1. Present the Links UI (pressing the links button on the toolbar).
1. Dismiss the Links UI bottom-sheet.
1. Press the hide-keyboard button on the toolbar.
1. Check that the button works and the keyboard is dismissed.
1. Check that the block is now deselected.

### Android

> **Note** On Android, the keyboard is already hidden after dismissing the Links UI.

1. Select a Paragraph block.
1. Press the hide-keyboard button on the toolbar.
1. Check that the keyboard is dismissed.
1. Check that the block is now deselected.

</details>

<details><summary>Verify focus + selection conflict resolution</summary>

1. Open a post in the editor.
1. Add a Paragraph block. 
1. Add an Image block beneath the Paragraph block and set its media. 
1. Simultaneously tap both the Image and Paragraph block to select the Image block, but focus the Paragraph block. (It can take a few attempts to accomplish this race condition bug.) 
1. Verify tapping the keyboard hide button in the block toolbar dismisses the keyboard.[^1]

[^1]: This works on iOS, but fails on Android. 

</details>

<details><summary>Test general Writing Flow behavior</summary>

Given the foundational nature of this change, it feels appropriate to thoroughly test the writing experience for the block editor in general. Common [Writing Flow tests are documented](https://github.com/wordpress-mobile/test-cases/tree/trunk/test-cases/gutenberg/writing-flow). 

</details>

## Screenshots or screencast

See:

- wordpress-mobile/WordPress-iOS#18783.
- https://github.com/wordpress-mobile/WordPress-iOS/issues/18783
